core/src/avm2/names.rs

```diff
@@ -0,0 +1,309 @@
+    }
+
+    pub fn package(package_name: &str) -> Self {
+        Namespace::Package(package_name.to_string())
+    }
+
+    pub fn is_any(&self) -> bool {
+        match self {
+            Self::Any => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_private(&self) -> bool {
+        match self {
+            Self::Private(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// A `QName`, likely "qualified name", consists of a namespace and name string.
+///
+/// This is technically interchangeable with `xml::XMLName`, as they both
+/// implement `QName`; however, AVM2 and XML have separate representations.
+///
+/// A property cannot be retrieved or set without first being resolved into a
+/// `QName`. All other forms of names and multinames are either versions of
+/// `QName` with unspecified parameters, or multiple names to be checked in
+/// order.
+#[derive(Clone, Collect, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[collect(no_drop)]
+pub struct QName {
+    ns: Namespace,
+    name: String,
+}
+
+impl QName {
+    pub fn new(ns: Namespace, name: &str) -> Self {
+        Self {
+            ns,
+            name: name.to_string(),
+        }
+    }
+
+    pub fn dynamic_name(local_part: &str) -> Self {
+        Self {
+            ns: Namespace::public_namespace(),
+            name: local_part.to_string(),
+        }
+    }
+
+    /// Pull a `QName` from the multiname pool.
+    ///
+    /// This function returns an Err if the multiname does not exist or is not
+    /// a `QName`.
+    pub fn from_abc_multiname(
+        file: &AbcFile,
+        multiname_index: Index<AbcMultiname>,
+    ) -> Result<Self, Error> {
+        let actual_index = multiname_index.0 as usize - 1;
+        let abc_multiname: Result<&AbcMultiname, Error> = file
+            .constant_pool
+            .multinames
+            .get(actual_index)
+            .ok_or_else(|| format!("Unknown multiname constant {}", multiname_index.0).into());
+
+        Ok(match abc_multiname? {
+            AbcMultiname::QName { namespace, name } => Self {
+                ns: Namespace::from_abc_namespace(file, namespace.clone())?,
+                name: abc_string(file, name.clone())?,
+            },
+            _ => return Err("Attempted to pull QName from non-QName multiname".into()),
+        })
+    }
+
+    pub fn local_name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn namespace(&self) -> &Namespace {
+        &self.ns
+    }
+}
+
+/// A `Multiname` consists of a name which could be resolved in one or more
+/// potential namespaces.
+///
+/// All unresolved names are of the form `Multiname`, and the name resolution
+/// process consists of searching each name space for a given name.
+///
+/// The existence of a `name` of `None` indicates the `Any` name.
+#[derive(Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct Multiname {
+    ns: Vec<Namespace>,
+    name: Option<String>,
```

Comment:
Is it possible for these to be Cows? We seem to construct them frequently based on existing data

Reply:
Namespaces are now `AvmString`s, so I believe there is no further need for deduplication as a namespace is just a string reference and an enum tag now.